mailpoet/lib/Automation/Integrations/MailPoet/Actions/SendWelcomeEmail/Action.php

```diff
@@ -0,0 +1,134 @@
+
+  /** @var ScheduledTasksRepository */
+  private $scheduledTasksRepository;
+
+  /** @var SubscribersFinder */
+  private $subscribersFinder;
+
+  public function __construct(
+    WelcomeScheduler $welcomeScheduler,
+    NewslettersRepository $newslettersRepository,
+    ScheduledTasksRepository $scheduledTasksRepository,
+    SubscribersFinder $subscribersFinder
+  ) {
+    $this->welcomeScheduler = $welcomeScheduler;
+    $this->newslettersRepository = $newslettersRepository;
+    $this->scheduledTasksRepository = $scheduledTasksRepository;
+    $this->subscribersFinder = $subscribersFinder;
+  }
+
+  public function getKey(): string {
+    return 'mailpoet:send-welcome-email';
+  }
+
+  public function getName(): string {
+    return __('Send Welcome Email', 'mailpoet');
+  }
+
+  public function validate(Workflow $workflow, Step $step, array $subjects = []): ValidationResult {
+    $result = new ValidationResult();
+    if (!isset($step->getArgs()['welcomeEmailId'])) {
+      $result->addError('welcomeEmailIdRequired', 'Step arguments did not include a welcomeEmailId.');
+    } else {
+      $welcomeEmailId = (int)$step->getArgs()['welcomeEmailId'];
+      $newsletter = $this->newslettersRepository->findOneById($welcomeEmailId);
+      if ($newsletter === null) {
+        $result->addError('welcomeEmailNotFound', sprintf("Welcome Email with ID '%s' not found.", $welcomeEmailId));
+      } else {
+        $type = $newsletter->getType();
+        if ($type !== NewsletterEntity::TYPE_WELCOME) {
+          $result->addError('newsletterMustBeWelcomeType', sprintf("Newsletter must be a Welcome Email. Actual type for newsletter ID '%s' was '%s'.", $welcomeEmailId, $type));
+        } else {
+          $result->setNewsletter($newsletter);
+        }
+      }
+    }
+
+    $segmentSubject = $subjects['mailpoet:segment'] ?? null;
+    if (!$segmentSubject instanceof SegmentSubject) {
+      $result->addError('segmentSubjectRequired', "A 'mailpoet:segment' subject is required.");
+    } else {
+      $result->setSegmentSubject($segmentSubject);
+    }
+
+    $subscriberSubject = $subjects['mailpoet:subscriber'] ?? null;
+    if (!$subscriberSubject instanceof SubscriberSubject) {
+      $result->addError('subscriberSubjectRequired', "A 'mailpoet:subscriber' subject is required.");
+    } else {
+      $result->setSubscriberSubject($subscriberSubject);
+    }
+
+    return $result;
+  }
+
+  public function run(Workflow $workflow, WorkflowRun $workflowRun, Step $step): void {
+    $validationResult = $this->validate($workflow, $step, $workflowRun->getSubjects());
+
+    if (!$validationResult->isValid()) {
+      throw InvalidStateException::create()->withErrors($validationResult->getErrors());
+    }
+
+    $newsletter = $validationResult->getNewsletter();
+    $subscriber = $validationResult->getSubscriberSubject()->getSubscriber();
+    if (!$subscriber instanceof SubscriberEntity) {
+      throw InvalidStateException::create();
```

Comment:
I'll add specific error messages once we've settled on an approach. 

Will we want all error messages to be translatable strings? Or just the ones that can occur as part of the "static" validation process, i.e. ones that can be triggered in the UI?

Reply:
Good question. These (runtime) errors will be logged and probably displayable somewhere in the UI. So I guess in the WP world it means we should make them translatable.

Reply:
Do we want to make all of the strings translatable now, or would it be better to wait until this feature is in more of a final state? I imagine things could change, and I don't know how big of a deal it is to ask our translators to translate strings that we might not end up using.

If these messages are going to be visible in the UI, should we get them approved first and have the approved messages be part of the ticket?

Reply:
I'm not sure what's the process here. A couple of new strings should be OK but as far as the whole feature goes, we should probably let all the original strings have checked before the MVP goes out. I feel like we need to add some messages to the exceptions so it's clear for us, not necessarily translatable now, but then there's the risk way may forget to make them translatable later on.